src/engine/evaluate.ts

```diff
@@ -7,26 +7,39 @@ import {
   AudienceDefinitionFilter,
 } from '../../types';
 
+// Validate whether a set of pageViews passes a set of rules
 export const evaluateCondition = (
   condition: EngineCondition<AudienceDefinitionFilter>,
   pageViews: PageView[]
 ): boolean => {
   const { filter, rules } = condition;
 
+  // We could possibly construct a list of various results for given
+  // queries. For instance, if (this query is one we've encountered for
+  // this pageViews set) { return the prior result set without doing any iteration }.
+  // filteredPageViews = hasQuery(resultsSet) ? return resultsSet[id] :
+  // filter.queries...
   const filteredPageViews = filter.queries.flatMap((query) =>
     pageViews.filter((pageView) => {
       const pageFeatures = pageView.features[query.queryProperty];
       return queryMatches(query, pageFeatures);
     })
   );
 
-  return rules.every((rule) => {
-    // TODO: allow other reducers...
-    //     // const reducer = reducers[rule.reducer.name](rule.reducer.args);
+  // Instead of iterating through everything, assume everything matches
+  // first, and exit the loop early if we find anything that doesn't.
+  // Further: Can the rules array be sorted in some way? If so, we can
+  // possibly use a binary search for O(log n) performance.
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

`every` does short-circuit immediately on the finding of a falsy element.
Does this change provide performance improvements anyway?

Reply:
Oh, I hadn't realized that - I'll look into it and maybe make a performance test between the two. If that's the spec, then we might not get gains here. :thinking: